### DISCUSSION TITLE
Is the order of stress and strain components correct in AbaqusUMATStress?

### DISCUSSION THREAD

**Discussion by ngrilli** [idaholab/moose#20109] [Q&A Modules: Solid mechanics] [answered]
Dear MOOSE Team,
 
In AbaqusUMATStress.C the ordering of the stress and strain seems to be different from the Abaqus convention:
https://abaqus-docs.mit.edu/2017/English/SIMACAEMODRefMap/simamod-c-conventions.htm

In Abaqus the ordering of the stress vector is:
sigma_{xx}, sigma_{yy}, sigma_{zz}, sigma_{xy}, sigma_{xz}, sigma_{yz}

however, the way _stress[_qp] is assigned at line 283 of AbaqusUMATStress is different
because the RankTwoTensor constructor follow another convention where sigma_{yz} and  sigma_{xy}
are inverted.

I was investigating this problem because of convergence issues I was facing when using the UMAT plugin.
Basically the model converged well with one element but gave wrong results as soon as more elements were used.

This problem was reported by @EdwardXZ06 some time ago:
https://github.com/idaholab/moose/discussions/19574

but I jumped too quickly to the conclusion that the timestepper was the problem.

Now I tried changing line 174 with the strain components order:
{{0, 0}, {1, 1}, {2, 2}, {0, 1}, {0, 2},{1, 2} }};

and the stress at line 284 to:

_aqSTRESS[0], _aqSTRESS[1], _aqSTRESS[2], _aqSTRESS[5], _aqSTRESS[4], _aqSTRESS[3]);

and all the convergence problems disappeared and I can get the same results as in Abaqus.

Of course, if my UMAT is adapted one can use the current AbaqusUMATStress,
however, I think the idea here is to provide compatibility between Abaqus UMAT and MOOSE,
therefore it would be better to follow Abaqus convention here. 

Should AbaqusUMATStress be changed? Can you please check?
Thank you

Best Regards,
Nicolò

**Comment by matthiasneuner** [comment]
Hi, I am using Abaqus Umats through a wrapper (https://github.com/matthiasneuner/chamois/blob/master/src/materials/ConvertRankTwoTensorToVoigt.C), and I am using following ordering:

```
  _the_rank_two_tensor_in_voigt[_qp][0] = _the_rank_two_tensor[_qp]( 0, 0 );
  _the_rank_two_tensor_in_voigt[_qp][1] = _the_rank_two_tensor[_qp]( 1, 1 );
  _the_rank_two_tensor_in_voigt[_qp][2] = _the_rank_two_tensor[_qp]( 2, 2 );
  _the_rank_two_tensor_in_voigt[_qp][3] = _the_rank_two_tensor[_qp]( 0, 1 );
  _the_rank_two_tensor_in_voigt[_qp][4] = _the_rank_two_tensor[_qp]( 0, 2 );
  _the_rank_two_tensor_in_voigt[_qp][5] = _the_rank_two_tensor[_qp]( 1, 2 );

  if ( _multiply_shear_terms_x2 )
  {
    _the_rank_two_tensor_in_voigt[_qp][3] *= 2;
    _the_rank_two_tensor_in_voigt[_qp][4] *= 2;
    _the_rank_two_tensor_in_voigt[_qp][5] *= 2;
  }
```

This seems consistent with your workaround -- if that helps.

**Comment by recuero** [comment]
Thanks for bringing this up. Opened an issue: https://github.com/idaholab/moose/issues/20111.

Definitely, our intention is to provide compatibility and not add additional steps on the user side.

**Comment by ngrilli** [comment, marked as answer]
Dear @recuero 

Also, the order of the components of _aqDDSDDE
that are provided to fillSymmetric21FromInputVector
to give the _jacobian_mult
may be wrong for the same reason.
I will analyse this and let you know.

The reason why this bug cannot be noticed in the tests
is because on one element with pure tension BC there is no shear strain or stress
so you will not notice the difference.

Best Regards,
Nicolò

**Reply by ngrilli** [reply]
Thank you for opening the issue

**Reply by recuero** [reply]
You are right. The order of Jacobian entries coming from UMAT does not match MOOSE's convention for the same reasons. The changes in https://github.com/idaholab/moose/pull/20112 improve the Jacobian coming from UMAT for a problem where shear stiffness depends on the direction. I tested that on a one-element problem, but should work also for other problems (not sure what constitutive equations you are using).

**Reply by ngrilli** [reply]
@recuero  Thank you very much for this, at a first glance it looks correct.
I am using an anisotropic elasto-plastic model, therefore I will check with the new jacobian and I should be able to see some changes in the convergence rate.
I will let you know if there are further problems.
Best Regards,
Nicolò
